gax/src/query_parameter.rs

```diff
@@ -14,44 +14,91 @@
 
 type Result<T> = std::result::Result<T, crate::request_parameter::Error>;
 
-/// Formats a query parameter.
+/// Adds a query parameter to a builder.
 ///
 /// Google APIs use [gRPC Transcoding](https://google.aip.dev/127). Some request
 /// fields are sent as query parameters and may need special formatting:
-/// - [Option] fields that do not contain a value are not included in the HTTP
-///   query.
+/// - Simple scalars are formatted as usual.
 /// - Fields of well-known types are formatted as strings. These include
 ///   [Duration](types::Duration), [FieldMask](types::FieldMask), and
 ///   [Timestamp](types::Timestamp).
-/// - Simple scalars are formatted as usual.
+/// - [Option] fields that do not contain a value are not included in the HTTP
+///   query.
+/// - Repeated fields are formatted as repeated query parameters.
+/// - Object fields use `field.subfield` format, and may (but rarely do)
+///   recurse.
 ///
 /// This function is called from the generated code. It is not intended for
 /// general use. The goal  
-pub fn format<T>(name: &'static str, parameter: &T) -> Result<Option<(&'static str, String)>>
+pub fn add<T>(
+    builder: reqwest::RequestBuilder,
+    name: &str,
+    parameter: &T,
+) -> Result<reqwest::RequestBuilder>
 where
     T: QueryParameter,
 {
-    QueryParameter::format(parameter)
-        .map(|result| result.map(|s| (name, s)))
-        .transpose()
+    QueryParameter::add(parameter, builder, name)
 }
 
 /// [QueryParameter] is a trait representing types that can be used as a query
 /// parameter.
-///
 pub trait QueryParameter {
-    fn format(&self) -> Option<Result<String>>;
+    fn add(&self, builder: reqwest::RequestBuilder, name: &str) -> Result<reqwest::RequestBuilder>;
```

Comment:
Seeing the generated code makes me wonder if we should have a query builder type. The shape here is trying to be a method on RequestBuilder, but of course we can't really do that. Just a thought for the future...

Reply:
Ack.